### PR TITLE
Allow dependently-typed object method declarations and fix subtyping

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
This PR
- allows dependently-typed object method declarations. Any class or object constructor that references/includes the type has to have an `external` implementation. Fixes https://github.com/ballerina-platform/ballerina-lang/issues/26430.

```ballerina
type Foo object {
    public function get(typedesc<anydata> td) returns td|error;
};

class Bar {
    *Foo;

    public function get(typedesc<anydata> td) returns td|error = external;
}
```

- Fixes subtyping for objects with dependently-typed methods. Fixes https://github.com/ballerina-platform/ballerina-lang/issues/26584.

The following is now allowed.

```ballerina
public class Foo {
    public function get(typedesc<anydata> td) returns td|error = external;
}

public class Bar {
    public function get(typedesc<anydata> td) returns td|error = external;
}

public function main() {
    Foo f = new Bar();
}
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by pubudu91** [comment]
Shouldn't the following work?
```ballerina
public function main() {
    Person p = new Employee("Pubudu", "Fernando");
    string s = p.getFullName(string);
}

class Person {
    string fname;
    string lname;

    function init(string fname, string lname) {
        self.lname = lname;
        self.fname = fname;
    }

    function getFullName(typedesc<anydata> td)  returns td = external;
}

class Employee {
    string fname;
    string lname;

    function init(string fname, string lname) {
        self.lname = lname;
        self.fname = fname;
    }

    function getFullName(typedesc<any> dt)  returns int = external;
}
```

**Comment by MaryamZi** [comment]
> Shouldn't the following work?
> 
> ```ballerina
> public function main() {
>     Person p = new Employee("Pubudu", "Fernando");
>     string s = p.getFullName(string);
> }
> 
> class Person {
>     string fname;
>     string lname;
> 
>     function init(string fname, string lname) {
>         self.lname = lname;
>         self.fname = fname;
>     }
> 
>     function getFullName(typedesc<anydata> td)  returns td = external;
> }
> 
> class Employee {
>     string fname;
>     string lname;
> 
>     function init(string fname, string lname) {
>         self.lname = lname;
>         self.fname = fname;
>     }
> 
>     function getFullName(typedesc<any> dt)  returns int = external;
> }
> ```

If we allow this, at runtime when we call `p.getFullName(string)`, what gets called is `Employee.getFullName` which returns an `int`. So where the user expects a `string`, he will get an `int`, so this cannot be allowed.

https://github.com/ballerina-platform/ballerina-spec/issues/633 discusses most of the scenarios related to this.

**Review by pubudu91 — APPROVED** [review]
Tested the changes locally. LGTM.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/types/ParameterizedType.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package io.ballerina.runtime.api.types;
+
+/**
+ * {@code ParameterizedType} represents the parameterized type in dependently-typed functions.
+ *
+ * @since 2.0.0
+ */
+public interface ParameterizedType extends Type {
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/types/ParameterizedType.java:26]: Do we need to export creating ParameterizedType using Typecreator api for interop developers?

**Review by warunalakshitha — APPROVED** [review]

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/types/ParameterizedType.java (hunk 1 above)
Comment (MaryamZi) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/types/ParameterizedType.java:26]: Right now, I can't think of a scenario where there'll be a requirement to create one in the external implementation. 

`ParameterizedType` is only expected to be used as the return type of a function/method atm anyway. Given that we don't expose creating a `FunctionType` right now, as discussed I think we can defer this until there is a requirement.

**Pull request merged**